File: webapp/models/intent_customer_diagnosis.py
from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, Field

from webapp.models.customer import BaseMongoModel


class IntentCustomerMeterConfig(BaseModel):
    meter_id: str = Field(..., min_length=1, description="???")
    account_id: str = Field(..., min_length=1, description="??")
    extracted_customer_name: Optional[str] = Field(None, description="?????????")
    multiplier: float = Field(1.0, gt=0, description="??")
    source_filename: Optional[str] = Field(None, description="?????")


class IntentCustomerProfile(BaseMongoModel):
    customer_name: str = Field(..., min_length=1, description="??????")
    created_at: datetime = Field(default_factory=datetime.now, description="????")
    updated_at: datetime = Field(default_factory=datetime.now, description="????")
    last_imported_at: Optional[datetime] = Field(None, description="??????")
    last_aggregated_at: Optional[datetime] = Field(None, description="??????")
    coverage_start: Optional[str] = Field(None, description="??????")
    coverage_end: Optional[str] = Field(None, description="??????")
    coverage_days: int = Field(0, description="????")
    missing_days: int = Field(0, description="????")
    completeness: float = Field(0.0, description="???")
    avg_daily_load: float = Field(0.0, description="????")
    max_daily_load: float = Field(0.0, description="?????")
    min_daily_load: float = Field(0.0, description="?????")
    missing_meter_days: int = Field(0, description="??????")
    interpolated_days: int = Field(0, description="????")
    dirty_days: int = Field(0, description="?????")
    meter_count: int = Field(0, description="????")
    meters: List[IntentCustomerMeterConfig] = Field(default_factory=list, description="????")


class IntentCustomerListItem(BaseModel):
    id: str = Field(..., description="??ID")
    customer_name: str = Field(..., description="????")
    created_at: datetime = Field(..., description="????")
    updated_at: datetime = Field(..., description="????")
    last_imported_at: Optional[datetime] = Field(None, description="??????")
    last_aggregated_at: Optional[datetime] = Field(None, description="??????")
    coverage_start: Optional[str] = Field(None, description="??????")
    coverage_end: Optional[str] = Field(None, description="??????")
    coverage_days: int = Field(0, description="????")
    missing_days: int = Field(0, description="????")
    completeness: float = Field(0.0, description="???")
    avg_daily_load: float = Field(0.0, description="????")
    max_daily_load: float = Field(0.0, description="?????")
    min_daily_load: float = Field(0.0, description="?????")
    missing_meter_days: int = Field(0, description="??????")
    interpolated_days: int = Field(0, description="????")
    dirty_days: int = Field(0, description="?????")
    meter_count: int = Field(0, description="????")


class IntentCustomerListResponse(BaseModel):
    items: List[IntentCustomerListItem] = Field(default_factory=list, description="??????")


class PreviewFileItem(BaseModel):
    filename: str = Field(..., description="???")
    meter_id: str = Field(..., description="???")
    account_id: str = Field(..., description="??")
    extracted_customer_name: Optional[str] = Field(None, description="?????????")
    start_date: str = Field(..., description="????")
    end_date: str = Field(..., description="????")
    record_count: int = Field(..., description="?????")
    default_multiplier: float = Field(1.0, description="????")
    parse_errors: List[str] = Field(default_factory=list, description="????")


class PreviewValidationResult(BaseModel):
    can_import: bool = Field(False, description="??????")
    errors: List[str] = Field(default_factory=list, description="????")
    warnings: List[str] = Field(default_factory=list, description="????")


class PreviewResponse(BaseModel):
    suggested_customer_name: Optional[str] = Field(None, description="??????")
    files: List[PreviewFileItem] = Field(default_factory=list, description="?????")
    validation: PreviewValidationResult = Field(default_factory=PreviewValidationResult, description="????")


class ImportMeterConfig(BaseModel):
    filename: str = Field(..., min_length=1, description="???")
    meter_id: str = Field(..., min_length=1, description="???")
    account_id: str = Field(..., min_length=1, description="??")
    multiplier: float = Field(1.0, gt=0, description="??")


class ImportResult(BaseModel):
    customer: IntentCustomerListItem
    imported_days: int = Field(..., description="???????")
    aggregated_days: int = Field(..., description="????")
    files: int = Field(..., description="????")
    message: str = Field(..., description="????")


class LoadSummaryResponse(BaseModel):
    customer: IntentCustomerListItem
    month_data: List[Dict] = Field(default_factory=list, description="???????")
    intraday_data: List[Dict] = Field(default_factory=list, description="??48????")
    selected_day_total: float = Field(0.0, description="??????")


class DeleteIntentCustomerRequest(BaseModel):
    password: str = Field(..., min_length=1, description="????")


class IntentWholesaleMonthlySummaryRow(BaseModel):
    settlement_month: str = Field(..., description="????")
    total_energy_mwh: float = Field(0.0, description="???(MWh)")
    daily_cost_total: float = Field(0.0, description="??????(?)")
    daily_cost_unit_price: float = Field(0.0, description="??????(?/MWh)")
    surplus_unit_price: float = Field(0.0, description="????????(?/MWh)")
    surplus_cost: float = Field(0.0, description="??????(?)")
    total_cost: float = Field(0.0, description="?????(?)")
    unit_cost_yuan_per_mwh: float = Field(0.0, description="????(?/MWh)")
    unit_cost_yuan_per_kwh: float = Field(0.0, description="????(?/kWh)")
    status: str = Field("success", description="????")
    message: str = Field("", description="????")


class IntentWholesalePeriodDetail(BaseModel):
    period: int = Field(..., description="????")
    time_label: str = Field(..., description="????")
    load_mwh: float = Field(0.0, description="??(MWh)")
    daily_cost_total: float = Field(0.0, description="??????(?)")
    surplus_cost: float = Field(0.0, description="??????(?)")
    total_cost: float = Field(0.0, description="???(?)")
    period_type: str = Field("平段", description="????")
    daily_cost_unit_price: float = Field(0.0, description="??????(?/MWh)")
    final_unit_price: float = Field(0.0, description="??????(?/MWh)")


class IntentWholesaleDailyDetail(BaseModel):
    date: str = Field(..., description="??")
    total_energy_mwh: float = Field(0.0, description="???(MWh)")
    daily_cost_total: float = Field(0.0, description="??????(?)")
    surplus_cost: float = Field(0.0, description="??????(?)")
    total_cost: float = Field(0.0, description="???(?)")
    unit_cost_yuan_per_mwh: float = Field(0.0, description="????(?/MWh)")


class IntentWholesaleMonthlyDetail(BaseModel):
    settlement_month: str = Field(..., description="????")
    summary: IntentWholesaleMonthlySummaryRow = Field(..., description="????")
    period_details: List[IntentWholesalePeriodDetail] = Field(default_factory=list, description="48????")
    daily_details: List[IntentWholesaleDailyDetail] = Field(default_factory=list, description="????")


class IntentWholesaleSimulationResponse(BaseModel):
    customer: IntentCustomerListItem = Field(..., description="????")
    summary_rows: List[IntentWholesaleMonthlySummaryRow] = Field(default_factory=list, description="?????")
    month_details: List[IntentWholesaleMonthlyDetail] = Field(default_factory=list, description="????")


class IntentRetailPackageOption(BaseModel):
    package_id: str = Field(..., description="??ID")
    package_name: str = Field(..., description="????")
    package_type: Optional[str] = Field(None, description="????")
    model_code: Optional[str] = Field(None, description="????")
    is_green_power: bool = Field(False, description="??????")
    status: Optional[str] = Field(None, description="??")


class IntentRetailPackageOptionsResponse(BaseModel):
    items: List[IntentRetailPackageOption] = Field(default_factory=list, description="??????")


class IntentRetailCalculatedPackageItem(BaseModel):
    package_id: str = Field(..., description="??ID")
    package_name: str = Field(..., description="????")
    model_code: Optional[str] = Field(None, description="????")
    updated_at: Optional[datetime] = Field(None, description="??????")


class IntentRetailCalculatedPackagesResponse(BaseModel):
    items: List[IntentRetailCalculatedPackageItem] = Field(default_factory=list, description="???????")


class IntentDeleteResultResponse(BaseModel):
    deleted_count: int = Field(0, description="?????")
    message: str = Field("", description="????")


class IntentRetailMonthResultRow(BaseModel):
    settlement_month: str = Field(..., description="????")
    total_energy_mwh: float = Field(0.0, description="???(MWh)")
    wholesale_unit_price: float = Field(0.0, description="????(?/MWh)")
    wholesale_amount: float = Field(0.0, description="????(?)")
    retail_unit_price: float = Field(0.0, description="????(?/MWh)")
    retail_amount: float = Field(0.0, description="????(?)")
    monthly_gross_profit: float = Field(0.0, description="???(?)")
    price_spread_per_mwh: float = Field(0.0, description="????(?/MWh)")
    is_capped: bool = Field(False, description="????")


class IntentRetailMonthResultsResponse(BaseModel):
    customer: IntentCustomerListItem = Field(..., description="????")
    package_id: str = Field(..., description="??ID")
    package_name: str = Field(..., description="????")
    rows: List[IntentRetailMonthResultRow] = Field(default_factory=list, description="??????")


class IntentRetailSimulationRequest(BaseModel):
    package_id: str = Field(..., min_length=1, description="??ID")


class IntentRetailSimulationDetailResponse(BaseModel):
    customer_id: str = Field(..., description="??ID")
    customer_name: str = Field(..., description="????")
    settlement_month: str = Field(..., description="????")
    package_id: str = Field(..., description="??ID")
    package_name: str = Field(..., description="????")
    model_code: Optional[str] = Field(None, description="????")
    price_model: Dict = Field(default_factory=dict, description="??????")
    pre_stage: Dict = Field(default_factory=dict, description="?????")
    sttl_stage: Dict = Field(default_factory=dict, description="?????")
    refund_context: Dict = Field(default_factory=dict, description="?????")
    final_stage: Dict = Field(default_factory=dict, description="????")
    period_details: List[Dict] = Field(default_factory=list, description="48????")
    daily_details: List[Dict] = Field(default_factory=list, description="????")
    final_energy_mwh: float = Field(0.0, description="????")
    final_retail_fee: float = Field(0.0, description="??????")
    final_retail_unit_price: float = Field(0.0, description="??????")
    final_wholesale_fee: float = Field(0.0, description="??????")
    final_wholesale_unit_price: float = Field(0.0, description="??????")
    final_gross_profit: float = Field(0.0, description="????")
    final_price_spread_per_mwh: float = Field(0.0, description="??????")
    final_excess_refund_fee: float = Field(0.0, description="??????")
    sttl_balancing_energy_mwh: float = Field(0.0, description="????")
    sttl_balancing_retail_fee: float = Field(0.0, description="??????")
    sttl_balancing_wholesale_fee: float = Field(0.0, description="??????")

File: frontend/src/api/intentCustomerDiagnosis.ts
import apiClient from './client';

export interface IntentCustomerSummary {
    id: string;
    customer_name: string;
    created_at: string;
    updated_at: string;
    last_imported_at?: string | null;
    last_aggregated_at?: string | null;
    coverage_start?: string | null;
    coverage_end?: string | null;
    coverage_days: number;
    missing_days: number;
    completeness: number;
    avg_daily_load: number;
    max_daily_load: number;
    min_daily_load: number;
    missing_meter_days: number;
    interpolated_days: number;
    dirty_days: number;
    meter_count: number;
}

export interface IntentPreviewFileItem {
    filename: string;
    meter_id: string;
    account_id: string;
    extracted_customer_name?: string | null;
    start_date: string;
    end_date: string;
    record_count: number;
    default_multiplier: number;
    parse_errors: string[];
}

export interface IntentPreviewResponse {
    suggested_customer_name?: string | null;
    files: IntentPreviewFileItem[];
    validation: {
        can_import: boolean;
        errors: string[];
        warnings: string[];
    };
}

export interface IntentLoadDataResponse {
    customer: IntentCustomerSummary;
    month_data: Array<{ date: string; label: string; totalLoad: number; isMissing: boolean }>;
    intraday_data: Array<{ time: string; load: number }>;
    selected_day_total: number;
}

export interface IntentImportConfig {
    filename: string;
    meter_id: string;
    account_id: string;
    multiplier: number;
}

export interface IntentWholesaleSummaryRow {
    settlement_month: string;
    total_energy_mwh: number;
    daily_cost_total: number;
    daily_cost_unit_price: number;
    surplus_unit_price: number;
    surplus_cost: number;
    total_cost: number;
    unit_cost_yuan_per_mwh: number;
    unit_cost_yuan_per_kwh: number;
    status: string;
    message: string;
}

export interface IntentWholesalePeriodDetail {
    period: number;
    time_label: string;
    load_mwh: number;
    daily_cost_total: number;
    surplus_cost: number;
    total_cost: number;
    period_type?: string;
    daily_cost_unit_price: number;
    final_unit_price: number;
}

export interface IntentWholesaleDailyDetail {
    date: string;
    total_energy_mwh: number;
    daily_cost_total: number;
    surplus_cost: number;
    total_cost: number;
    unit_cost_yuan_per_mwh: number;
}

export interface IntentWholesaleMonthDetail {
    settlement_month: string;
    summary: IntentWholesaleSummaryRow;
    period_details: IntentWholesalePeriodDetail[];
    daily_details: IntentWholesaleDailyDetail[];
}

export interface IntentWholesaleSimulationResponse {
    customer: IntentCustomerSummary;
    summary_rows: IntentWholesaleSummaryRow[];
    month_details: IntentWholesaleMonthDetail[];
}

export interface IntentRetailPackageOption {
    package_id: string;
    package_name: string;
    package_type?: string | null;
    model_code?: string | null;
    is_green_power: boolean;
    status?: string | null;
}

export interface IntentRetailCalculatedPackageItem {
    package_id: string;
    package_name: string;
    model_code?: string | null;
    updated_at?: string | null;
}

export interface IntentDeleteResultResponse {
    deleted_count: number;
    message: string;
}

export interface IntentRetailMonthResultRow {
    settlement_month: string;
    total_energy_mwh: number;
    wholesale_unit_price: number;
    wholesale_amount: number;
    retail_unit_price: number;
    retail_amount: number;
    monthly_gross_profit: number;
    price_spread_per_mwh: number;
    is_capped: boolean;
}

export interface IntentRetailMonthResultsResponse {
    customer: IntentCustomerSummary;
    package_id: string;
    package_name: string;
    rows: IntentRetailMonthResultRow[];
}

export interface IntentRetailSimulationDetail {
    customer_id: string;
    customer_name: string;
    settlement_month: string;
    package_id: string;
    package_name: string;
    model_code?: string | null;
    price_model: Record<string, any>;
    pre_stage: Record<string, any>;
    sttl_stage: Record<string, any>;
    refund_context: Record<string, any>;
    final_stage: Record<string, any>;
    period_details: Array<Record<string, any>>;
    daily_details: Array<Record<string, any>>;
    final_energy_mwh: number;
    final_retail_fee: number;
    final_retail_unit_price: number;
    final_wholesale_fee: number;
    final_wholesale_unit_price: number;
    final_gross_profit: number;
    final_price_spread_per_mwh: number;
    final_excess_refund_fee: number;
    sttl_balancing_energy_mwh: number;
    sttl_balancing_retail_fee: number;
    sttl_balancing_wholesale_fee: number;
    pre_energy_mwh?: number;
    pre_retail_fee?: number;
    pre_retail_unit_price?: number;
    pre_wholesale_fee?: number;
    pre_wholesale_unit_price?: number;
    pre_gross_profit?: number;
    pre_price_spread_per_mwh?: number;
    sttl_energy_mwh?: number;
    sttl_retail_fee?: number;
    sttl_retail_unit_price?: number;
    sttl_wholesale_fee?: number;
    sttl_wholesale_unit_price?: number;
    sttl_gross_profit?: number;
    sttl_price_spread_per_mwh?: number;
}

export const listIntentCustomers = async () => {
    const response = await apiClient.get<{ items: IntentCustomerSummary[] }>('/api/v1/intent-customer-diagnosis/customers');
    return response.data.items;
};

export const previewIntentCustomerFiles = async (files: File[]) => {
    const formData = new FormData();
    files.forEach((file) => formData.append('files', file));
    const response = await apiClient.post<IntentPreviewResponse>('/api/v1/intent-customer-diagnosis/preview', formData, {
        headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
};

export const importIntentCustomerFiles = async (
    configs: IntentImportConfig[],
    files: File[]
) => {
    const formData = new FormData();
    formData.append('meter_configs_json', JSON.stringify(configs));
    files.forEach((file) => formData.append('files', file));
    const response = await apiClient.post('/api/v1/intent-customer-diagnosis/import', formData, {
        headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
};

export const getIntentCustomerLoadData = async (customerId: string, month: string, date: string) => {
    const response = await apiClient.get<IntentLoadDataResponse>(
        `/api/v1/intent-customer-diagnosis/customers/${customerId}/load-data`,
        { params: { month, date } }
    );
    return response.data;
};

export const deleteIntentCustomer = async (customerId: string, password: string) => {
    await apiClient.delete(`/api/v1/intent-customer-diagnosis/customers/${customerId}`, {
        data: { password }
    });
};

export const calculateIntentCustomerWholesaleSimulation = async (customerId: string) => {
    const response = await apiClient.post<IntentWholesaleSimulationResponse>(
        `/api/v1/intent-customer-diagnosis/customers/${customerId}/wholesale-simulation`
    );
    return response.data;
};

export const getIntentCustomerWholesaleSimulation = async (customerId: string) => {
    const response = await apiClient.get<IntentWholesaleSimulationResponse>(
        `/api/v1/intent-customer-diagnosis/customers/${customerId}/wholesale-simulation`
    );
    return response.data;
};

export const listIntentRetailActivePackages = async () => {
    const response = await apiClient.get<{ items: IntentRetailPackageOption[] }>(
        '/api/v1/intent-customer-diagnosis/retail-simulation/packages/active'
    );
    return response.data.items;
};

export const listIntentRetailCalculatedPackages = async (customerId: string) => {
    const response = await apiClient.get<{ items: IntentRetailCalculatedPackageItem[] }>(
        `/api/v1/intent-customer-diagnosis/customers/${customerId}/retail-simulation/packages`
    );
    return response.data.items;
};

export const listIntentRetailMonthResults = async (customerId: string, packageId: string) => {
    const response = await apiClient.get<IntentRetailMonthResultsResponse>(
        `/api/v1/intent-customer-diagnosis/customers/${customerId}/retail-simulation/months`,
        { params: { package_id: packageId } }
    );
    return response.data;
};

export const calculateIntentRetailSimulation = async (
    customerId: string,
    packageId: string
) => {
    const response = await apiClient.post<IntentRetailSimulationDetail>(
        `/api/v1/intent-customer-diagnosis/customers/${customerId}/retail-simulation`,
        {
            package_id: packageId
        }
    );
    return response.data;
};

export const getIntentRetailSimulationDetail = async (
    customerId: string,
    packageId: string,
    settlementMonth: string
) => {
    const response = await apiClient.get<IntentRetailSimulationDetail>(
        `/api/v1/intent-customer-diagnosis/customers/${customerId}/retail-simulation/detail`,
        { params: { package_id: packageId, settlement_month: settlementMonth } }
    );
    return response.data;
};

export const deleteIntentRetailSimulationPackage = async (
    customerId: string,
    packageId: string
) => {
    const response = await apiClient.delete<IntentDeleteResultResponse>(
        `/api/v1/intent-customer-diagnosis/customers/${customerId}/retail-simulation/packages/${packageId}`
    );
    return response.data;
};
